### PR TITLE
feat(ingestion): PowerBI# Improve PowerBI source ingestion

### PR DESCRIPTION
**Pull request by mohdsiddique**
**Improved mapping would be:**
    1) PowerBI DataSet's name -> DataHub DataSet db-name (It would be part of URN: <db-name>.<table-name>)
    2) PowerBI DataSet's table-name -> DataHub DataSet table-name (It would be part of URN: <db-name>.<table-name>)
    3) PowerBI DataSet's Platform i.e `powerbi` -> DataHub DataSet DataPlatform i.e `powerbi`

    <schema-name> wouldn't be available as PoweBI DataSet doesn't have a schema concept.

### CODE REVIEW

File: metadata-ingestion/docs/sources/powerbi/powerbi_pre.md [hunk 1]
```diff
@@ -7,14 +7,58 @@ See the
...
+| `Dataset's Table`     | `Dataset`           |
 | `Tile`                | `Chart`             |
 | `Report.webUrl`       | `Chart.externalUrl` |
 | `Workspace`           | `N/A`               |
 | `Report`              | `Dashboard`         |
 | `Page`                | `Chart`             |
 
 If Tile is created from report then Chart.externalUrl is set to Report.webUrl.
+
+## Lineage
```
Comment (jjoyce0510) [inline metadata-ingestion/docs/sources/powerbi/powerbi_pre.md:22]: Add one sentence here to descibe what type of lineage this source extracts

Comment (jjoyce0510) [reply metadata-ingestion/docs/sources/powerbi/powerbi_pre.md:22]: (Introduction sentence)

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py [hunk 2]
```diff
@@ -0,0 +1,911 @@
...
+        return "{}{}".format(PowerBiAPI.AUTHORITY, self.__config.tenant_id)
+
+    def __get_users(self, workspace_id: str, entity: str, _id: str) -> List[User]:
+        """
+        Get user for the given PowerBi entity
+        """
+        users: List[PowerBiAPI.User] = []
+        if self.__config.extract_ownership is False:
+            LOGGER.info(
+                "ExtractOwnership capabilities is disabled from configuration and hence returning empty users list"
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:225]: Extracting ownership is disabled from configuration... and so on

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py [hunk 3]
```diff
@@ -0,0 +1,911 @@
...
+                emailAddress=instance.get("emailAddress"),
+                graphId=instance.get("graphId"),
+                principalType=instance.get("principalType"),
+            )
+            for instance in users_dict
+        ]
+
+        return users
+
+    def __get_report(
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:272]: do we need 2 underscores here? in these method names. do we do this in other places?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py [hunk 4]
```diff
@@ -0,0 +1,911 @@
...
+        report_get_endpoint: str = PowerBiAPI.API_ENDPOINTS[Constant.REPORT_GET]
+        # Replace place holders
+        report_get_endpoint = report_get_endpoint.format(
+            POWERBI_BASE_URL=PowerBiAPI.BASE_URL,
+            WORKSPACE_ID=workspace_id,
+            REPORT_ID=report_id,
+        )
+        # Hit PowerBi
+        LOGGER.info(f"Request to report URL={report_get_endpoint}")
+        response = requests.get(
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:293]: minor: should we add retry logic? 

is there any scenario in which this method throws an exception? perhaps we should try: except on these external calls

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py [hunk 5]
```diff
@@ -0,0 +1,911 @@
...
+
+        data_sources: Dict[str, "PowerBiAPI.DataSource"] = {}
+        LOGGER.debug("data-sources = {}".format(value))
+        for datasource_dict in value:
+            # Create datasource instance with basic detail available
+            datasource = PowerBiAPI.DataSource(
+                id=datasource_dict.get(
+                    "datasourceId"
+                ),  # datasourceId is not available in all cases
+                type=datasource_dict["datasourceType"],
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:501]: are these fields guaranteed to be present?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py [hunk 6]
```diff
@@ -0,0 +1,911 @@
...
+            LOGGER.warning("{}={}".format(Constant.DatasetId, dataset.id))
+            LOGGER.warning("{}={}".format(Constant.HTTP_RESPONSE_TEXT, response.text))
+            LOGGER.warning(
+                "{}={}".format(Constant.HTTP_RESPONSE_STATUS_CODE, response.status_code)
+            )
+
+            raise ConnectionError(message)
+
+        res = response.json()
+        value = res["value"]
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:485]: is this guaranteed to be present?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py [hunk 7]
```diff
@@ -0,0 +1,911 @@
...
+            LOGGER.info(f"Max trial {max_trial}")
+            scan_get_endpoint = PowerBiAPI.API_ENDPOINTS[Constant.SCAN_GET]
+            scan_get_endpoint = scan_get_endpoint.format(
+                POWERBI_ADMIN_BASE_URL=PowerBiAPI.ADMIN_BASE_URL, SCAN_ID=scan_id
+            )
+
+            LOGGER.info(f"Hitting URL={scan_get_endpoint}")
+
+            trail = 1
+            while True:
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:776]: this is getting kinda difficult to follow. i'd recommend trying to split up this method..

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py [hunk 8]
```diff
@@ -0,0 +1,911 @@
...
+                LOGGER.info(f"Trial = {trail}")
+                res = requests.get(
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:778]: any reason to wrap this in another method that can try : except and potentially issue a retry as well?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py [hunk 9]
```diff
@@ -0,0 +1,911 @@
...
+
+                    raise ConnectionError(message)
+
+                if res.json()["status"].upper() == "Succeeded".upper():
+                    LOGGER.info(f"Scan result is available for scan id({scan_id})")
+                    return True
+
+                if trail == max_trial:
+                    break
+                LOGGER.info(f"Sleeping for {minimum_sleep} seconds")
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:795]: Do we usually use this LOGGER convention? I don't recall seeing this.

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py [hunk 10]
```diff
@@ -0,0 +1,911 @@
...
+                )
+                dataset_map[dataset_instance.id] = dataset_instance
+                # set dataset-name
+                dataset_name: str = (
+                    dataset_instance.name
+                    if dataset_instance.name is not None
+                    else dataset_instance.id
+                )
+
+                for table in dataset_dict["tables"]:
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:854]: is tableas guaranteed to be present?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py [hunk 11]
```diff
@@ -0,0 +1,911 @@
...
+                    expression: str = (
+                        table["source"][0]["expression"]
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:856]: is the first element in the list guaranteed to be present?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py [hunk 12]
```diff
@@ -0,0 +1,911 @@
...
+                        if table.get("source") is not None and len(table["source"]) > 0
+                        else None
+                    )
+                    dataset_instance.tables.append(
+                        PowerBiAPI.Table(
+                            name=table["name"],
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:862]: is name guaranteed to be present?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py [hunk 13]
```diff
@@ -0,0 +1,814 @@
...
+            return dataset_mcps
+
+        LOGGER.info(
+            f"Converting dataset={dataset.name}(id={dataset.id}) to datahub dataset"
+        )
+
+        for table in dataset.tables:
+            # Create a URN for dataset
+            ds_urn = builder.make_dataset_urn(
+                platform=self.__config.platform_name,
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:148]: Where is the platform instance? Is that supported in this source?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py [hunk 14]
```diff
@@ -0,0 +1,814 @@
...
+            # Remove status mcp
+            status_mcp = self.new_mcp(
+                entity_type=Constant.DATASET,
+                entity_urn=ds_urn,
+                aspect_name=Constant.STATUS,
+                aspect=StatusClass(removed=False),
+            )
+            dataset_mcps.extend([info_mcp, status_mcp])
+
+            if self.__config.extract_lineage is True:
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:175]: Starting at this line this methods get a bit too long. Can we split into more reasonably sized functions?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/resolver.py [hunk 15]
```diff
@@ -0,0 +1,575 @@
...
+            )
+            # delete arg_list as we consumed it and don't want to process it in next step
+            if len(arguments) != 2:
+                LOGGER.debug("Expected 2 arguments, but got {%s}", len(arguments))
+                return full_table_names
+
+            del token_dict[data_access_function]["arg_list"]
+
+            db_name: str = arguments[1]
+            for source in token_dict[data_access_function]:
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/resolver.py:327]: This code gets a bit complex

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/resolver.py [hunk 16]
```diff
@@ -0,0 +1,575 @@
...
+                source_dict: Dict[str, Any] = token_dict[data_access_function][source]
+
+                db_name: Optional[str] = self._get_db_name(arguments[0])
+                if db_name is None:
+                    return full_table_names
+
+                for schema in source_dict["item_selectors"]:
+                    schema_name: str = schema["items"]["Schema"]
+                    for item_selectors in source_dict[schema["assigned_to"]]:
+                        for item_selector in source_dict[schema["assigned_to"]][
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/resolver.py:425]: 3 nested for loops - any way to reduce the nesting here? it's very hard to follow nesting in python.

File: metadata-ingestion/docs/sources/powerbi/powerbi_pre.md [hunk 17]
```diff
@@ -7,14 +7,58 @@ See the
...
+You can control table lineage ingestion using `extract_lineage` configuration parameter, by default it is set to `true`. 
+
+PowerBI Source extracts the lineage information by parsing PowerBI M-Query expression.
+
+PowerBI Source supports M-Query expression for below listed PowerBI Data Sources 
+
+1.  Snowflake 
+2.  Oracle 
+3.  PostgreSQL
+4.  MS-SQL
```
Comment (jjoyce0510) [inline metadata-ingestion/docs/sources/powerbi/powerbi_pre.md:32]: Let's say "Microsoft SQL Server"

Comment (mohdsiddique) [reply metadata-ingestion/docs/sources/powerbi/powerbi_pre.md:32]: Done

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/resolver.py (hunk 16 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/resolver.py]: I resolved it, now we don't have loops

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/resolver.py (hunk 15 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/resolver.py]: Now, I made it more redable

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py (hunk 13 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:148]: platform instance is not supported in this source

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py (hunk 12 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:862]: It is guaranteed to be present.

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py (hunk 11 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:856]: As per document it is guaranteed to be present 
https://learn.microsoft.com/en-us/rest/api/power-bi/admin/workspace-info-get-scan-result#table

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py (hunk 10 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:854]: Yes

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py (hunk 9 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:795]: Yes, we do have this in oracle.py and snowflake_v2.py. The difference is I used the all-uppercase LOGGER instead of logger.

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py (hunk 8 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:778]: _create_scan_job_ is creating a job and _wait_for_scan_to_complete_ is waiting for a job to be completed. so no need to wrap it in another try : except

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py (hunk 6 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:485]: Yes

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py (hunk 5 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:501]: Yes

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py (hunk 4 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:293]: It only fails if the server is not reachable, in such cases user can re-run the ingestion.

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py (hunk 3 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:272]: Updated it to __get_report_

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py (hunk 2 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:225]: Done

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py (hunk 14 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:235]: Moved it to a separate function.

File: metadata-ingestion/docs/sources/powerbi/powerbi_pre.md (hunk 1 above)
Comment (mohdsiddique) [reply metadata-ingestion/docs/sources/powerbi/powerbi_pre.md:22]: Done

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py (hunk 7 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:776]: Added new function to calculate trial

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/parser.py [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+
+
+def _parse_expression(expression: str) -> Tree:
+    # Read lexical grammar as text
+    grammar: str = pkg_resource.read_text(
+        "datahub.ingestion.source.powerbi", "powerbi-lexical-grammar.rule"
+    )
+
+    # Create lark parser for the grammar text
+    lark_parser = Lark(grammar, start="let_expression", regex=True)
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/parser.py:22]: Can we save this as an higher level global variable? My guess is that creating the parser is not a cheap process?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/parser.py [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+
+    parse_tree: Tree = lark_parser.parse(expression)
+
+    LOGGER.debug("Parse Tree")
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/parser.py:26]: not super useful... Parsing expression {expression} would be more useful

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/resolver.py [hunk 20]
```diff
@@ -0,0 +1,655 @@
...
+                        data_platform_pair=table_full_name_creator.get_platform_pair(),
+                    )
+                )
+
+        return data_platform_tables
+
+
+class DefaultTwoStepDataAccessSources(AbstractTableFullNameCreator, ABC):
+    """
+    These are the DataSource for which PowerBI Desktop generates default M-Query of following pattern
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/resolver.py:373]: thanks for adding this comment

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/resolver.py [hunk 21]
```diff
@@ -0,0 +1,655 @@
...
+            IdentifierAccessor, data_access_func_detail.identifier_accessor
+        ).items["Schema"]
+
+        table_name: str = cast(
+            IdentifierAccessor, data_access_func_detail.identifier_accessor
+        ).items["Item"]
+
+        full_table_names.append(f"{db_name}.{schema_name}.{table_name}")
+
+        LOGGER.debug("PostgreSQL full-table-names = %s", full_table_names)
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/resolver.py:413]: always postgres?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/validator.py [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+from typing import List, Optional, Tuple
+
+from lark import Tree
+
+from datahub.ingestion.source.powerbi.m_query import resolver, tree_function
+
+LOGGER = logging.getLogger(__name__)
+
+
+def any_one_should_present(
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/validator.py:11]: any_function_should_be_known -> function should be named consistently with the one below.

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py [hunk 23]
```diff
@@ -0,0 +1,912 @@
...
+        LOGGER.info(f"Request to URL={dashboard_list_endpoint}")
+        response = requests.get(
+            dashboard_list_endpoint,
+            headers={Constant.Authorization: self.get_access_token()},
+        )
+
+        # Check if we got response from PowerBi
+        if response.status_code != 200:
+            LOGGER.warning("Failed to fetch dashboard list from power-bi for")
+            LOGGER.warning(f"{Constant.WorkspaceId}={workspace.id}")
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:378]: assuming much of this logic is copied from the original powerbi file

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py (hunk 9 above)
Comment (jjoyce0510) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:796]: why?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py (hunk 11 above)
Comment (jjoyce0510) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:857]: we should not trust them :)

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py (hunk 12 above)
Comment (jjoyce0510) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:863]: never trust the foreign APIs, unless we've validated the object ourselves. If Msft docs are lying, or more likely omitting some edge cases, we can incur runtime exceptions if we are not careful.

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py (hunk 9 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:796]: No specific reason. I named it in All caps as per python global constant should be in capital letters. I converted it to all lowercase

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/parser.py (hunk 18 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/parser.py:22]: Done

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/parser.py (hunk 19 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/parser.py:26]: done

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/resolver.py (hunk 21 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/resolver.py:413]: corrected

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py (hunk 23 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:378]: Yes complete code, for current work we haven't added any API.

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/validator.py (hunk 22 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/m_query/validator.py:11]: I removed these functions, they were not utilized.
